google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/SequentialExecutorService.java

```diff
@@ -0,0 +1,227 @@
+package com.google.cloud.pubsub.v1;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.SettableApiFuture;
+import com.google.pubsub.v1.PublishResponse;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.Executor;
+
+interface CancellableRunnable extends Runnable {
+  public void cancel(Throwable e);
+}
+
+/**
+ * An executor service that runs the tasks with the same key sequentially. The tasks with the same
+ * key will be run only when its predecessor has been completed while tasks with different keys can
+ * be run in parallel.
+ */
+final class SequentialExecutorService<T> {
+  private final SequentialExecutor manageableSequentialExecutor;
+  private final SequentialExecutor autoSequentialExecutor;
+
+  public SequentialExecutorService(Executor executor) {
+    this.manageableSequentialExecutor = SequentialExecutor.newManageableSequentialExecutor(executor);
+    this.autoSequentialExecutor = SequentialExecutor.newAutoSequentialExecutor(executor);
+  }
+
+  /**
+   * Runs asynchronous {@code Callable} tasks sequentially. If one of the tasks fails, other tasks
+   * with the same key that have not been executed will be cancelled.
+   */
+  public ApiFuture<T> submit(final String key, final Callable<ApiFuture> callable) {
+    final SettableApiFuture<T> future = SettableApiFuture.<T>create();
+    manageableSequentialExecutor.execute(key, new CancellableRunnable() {
+      private boolean cancelled = false;
+
+      @Override
+      public void run() {
+        if (cancelled) {
+          return;
+        }
+        try {
+          ApiFuture<T> callResult = callable.call();
+          ApiFutures.addCallback(callResult, new ApiFutureCallback<T>() {
+            @Override
+            public void onSuccess(T msg) {
+              future.set(msg);
+              manageableSequentialExecutor.resume(key);
+            }
+
+            @Override
+            public void onFailure(Throwable e) {
+              future.setException(e);
+              manageableSequentialExecutor.cancelQueuedTasks(key,
+                  new CancellationException(
+                      "Publishing cancelled since delivering previous message has been failed."));
```

Comment:
is it possible to add info about which previous message failed into the error message for easier debugging?

Reply:
change to "previous runnable" instead of message